runtime/src/proofs.rs

```diff
@@ -0,0 +1,282 @@
+use rstd::vec::Vec;
+
+#[derive(Debug)]
+pub struct Proof {
+    hash: [u8; 32],
+    sorted_hashes: Vec<[u8; 32]>,
+}
+
+/// validates each proof and return true if all the proofs are valid
+/// else returns false
+pub fn validate_proofs(doc_root: [u8; 32], proofs: &Vec<Proof>) -> bool {
+    if proofs.len() < 1 {
+        return false;
+    }
+
+    let mut matches = Vec::new();
+    matches.push(doc_root);
+    return proofs
+        .iter()
+        .map(|proof| {
+            validate_proof(
+                &mut matches,
+                proof.hash.clone(),
+                proof.sorted_hashes.clone(),
+            )
+        })
+        .fold(true, |acc, b| acc && b);
+}
+
+fn hash_of(a: [u8; 32], b: [u8; 32]) -> [u8; 32] {
+    let mut h: Vec<u8> = Vec::with_capacity(64);
+    if a < b {
+        h.extend_from_slice(&a);
+        h.extend_from_slice(&b);
+    } else {
+        h.extend_from_slice(&b);
+        h.extend_from_slice(&a);
+    }
+
+    runtime_io::blake2_256(&h)
+}
+
+fn validate_proof(matches: &mut Vec<[u8; 32]>, hash: [u8; 32], proofs: Vec<[u8; 32]>) -> bool {
+    // if hash is already cached earlier
+    if matches.contains(&hash) {
+        return true;
+    }
+
+    let mut hash = hash.clone();
+    for proof in proofs.into_iter() {
+        matches.push(proof);
+        hash = hash_of(hash, proof);
+        if matches.contains(&hash) {
+            return true;
+        }
+        matches.push(hash.clone())
+    }
+
+    false
+}
+
+// appends deposit_address and all the hashes from the proofs and returns keccak hash of the result.
+pub fn bundled_hash(proofs: Vec<Proof>, deposit_address: [u8; 20]) -> [u8; 32] {
+    let hash = proofs
+        .into_iter()
+        .fold(deposit_address.to_vec(), |mut acc, proof: Proof| {
```

Comment:
`deposit_addr+proofs` is the order? or `proofs+deposit_addr`?

Reply:
Deposit_address+Proofs as per your comment on the Asset contract PR